src/librustc_save_analysis/dump_visitor.rs

```diff
@@ -139,6 +145,138 @@ impl<'l, 'tcx: 'l, 'll, D: Dump + 'll> DumpVisitor<'l, 'tcx, 'll, D> {
+        let borrow_analysis = self.save_ctxt.borrow_analysis_map.drain().collect::<Vec<_>>();
+        for (def_id, analysis_result) in borrow_analysis {
+            self.process_borrow_analysis(def_id, analysis_result);
+        }
+    }
+
+    fn process_borrow_analysis(&mut self, def_id: DefId, mut analysis_result: AnalysisResult<'l, 'tcx>) {
+        let mut safe_loans = {
+            let mut safe_loans = vec![];
+            mem::swap(&mut safe_loans, &mut analysis_result.safe_loans);
+            safe_loans.into_iter()
+                .filter_map(|safe_loan| self.process_safe_loan(safe_loan))
+                .collect::<Vec<_>>()
+        };
+
+        let mut loans = {
+            let mut loans = vec![];
+            mem::swap(&mut loans, &mut analysis_result.all_loans);
+            loans.into_iter()
+                .filter_map(|loan| self.process_loan(def_id, loan))
+                .collect::<Vec<_>>()
+        };
+
+        loans.append(&mut safe_loans);
+
+        let moves = {
+            let mut moves = vec![];
+            mem::swap(&mut moves, &mut analysis_result.move_data.moves.borrow_mut());
+            moves.into_iter()
+                .map(|m| self.process_move(m, &analysis_result.move_data))
+                .collect::<Vec<_>>()
+        };
+
+        let scopes = {
+            let move_map: HashMap<Id, &MoveData> = moves.iter()
+                .fold(HashMap::new(), |mut acc, m| {
+                    match acc.entry(m.ref_id) {
+                        // we just want the 
+                        Entry::Occupied(mut e) => if e.get().span.byte_start > m.span.byte_start { e.insert(m); },
+                        Entry::Vacant(e) => { e.insert(m); },
+                    }
+
+                    acc
+                });
+            let mut assignments = vec![];
+            mem::swap(&mut assignments, &mut analysis_result.move_data.var_assignments.borrow_mut());
+            assignments.into_iter()
+                .filter_map(|a| {
+                    let id = ::id_from_node_id(a.assignee_id, &self.save_ctxt);
+                    let mov = move_map.get(&id);
+                    self.process_assignment(a, mov, &analysis_result.move_data, &analysis_result.bccx)
+                })
+                .collect()
+        };
+
+        let data = BorrowData {
+            ref_id: ::id_from_def_id(def_id),
+            scopes: scopes,
+            loans: loans,
+            moves: moves,
+        };
+
+        self.dumper.dump_per_fn_borrow_data(data);
+    }
+
+    fn process_assignment(
+        &self,
+        assignment: Assignment,
+        mov: Option<&&MoveData>,
+        move_data: &rustc_borrowck::MoveData<'tcx>,
+        bccx: &BorrowckCtxt<'l, 'tcx>)
+    -> Option<Scope> {
+        move_data.path_loan_path(assignment.path)
+            .kill_scope(&bccx)
+            .span(&self.tcx.hir)
+            .map(|span| {
+                let expanded_span = match mov.map(|m| &m.span) {
```

Comment:
Here I'm using the `kill_scope` of an assignment to determine the "scope" of a value. I shorten it to the end of the soonest related `move` if it's sooner to reflect what the compiler is doing internally for the non-mir case. This was something I clarified with @pnkfelix. Post-MIR borrowck complete, this will most likely have to be moved to within the old borrowck.